api/src/main/java/org/apache/iceberg/stats/StatsUtil.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.stats;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types;
+
+public class StatsUtil {
+  private static final int NUM_STATS_PER_COLUMN = 200;
+  static final int RESERVED_FIELD_IDS = 200;
+  static final int DATA_SPACE_FIELD_ID_START = 10_000;
+  static final int METADATA_SPACE_FIELD_ID_START = 1_417_000_000;
+  static final int RESERVED_FIELD_IDS_START = Integer.MAX_VALUE - RESERVED_FIELD_IDS;
+
+  private StatsUtil() {}
+
+  public static int statsFieldIdFor(int fieldId) {
```

Comment:
nit on the name, statsIdForField?

I may be being paranoid here but I feel like we are putting a lot of math in here and at least to me it's obscuring things a bit.

Maybe instead we could do

statsIdForField (fieldId) {
   Precondition (fieldID >= 0)
   return (fieldId >= RESEREVED_FIELD_ID_START) ? 
     statsIdForReservedField(fieldId) :
     statsIdForDataField(fieldId)
  }
  
  statsIdForDataField(fieldId) {
     statsId =  DATA_SPACE_FIELD_ID_START + NUM_STATS_PER_COLUMN * (long) id;
     //overflow conditions specific to data field
  }
  
  statsIdForReservedField(fieldId) {
     statId = RESERVED_FIELD_IDS - (Integer.MAX_VALUE - fieldID)
     // Overflow check for ReservedFields
  }